plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/FlowHead.java

```diff
@@ -151,17 +143,31 @@ void setNewHead(@NonNull FlowNode v) {
             LOGGER.log(Level.WARNING, "Failed to record new head or persist old: " + v, e);
         }
         this.head = v;
-        CpsThreadGroup c = CpsThreadGroup.current();
-        if (c !=null) {
-            // if the manipulation is from within the program executing thread, then
-            // defer the notification till we get to a safe point.
-            c.notifyNewHead(v);
+        notifyNewHead(v, asynchNotifications);
+    }
+
+    /**
+     * Usually calls {@link CpsThreadGroup#notifyNewHead}.
+     * If the manipulation is from within the program executing thread,
+     * then defer the notification till we get to a safe point.
+     * Can also be used for special situations such as a fatal error;
+     * in those cases we may need to grow the graph without running the program,
+     * and may also want to avoid holding any locks.
+     */
+    private void notifyNewHead(@NonNull FlowNode head, boolean asynchIfOutsideProgram) {
+        var g = CpsThreadGroup.current();
+        if (g !=null) {
+            g.notifyNewHead(head);
         } else {
-            // in recovering from error and such situation, we sometimes need to grow the graph
-            // without running the program.
-            // TODO can CpsThreadGroup.notifyNewHead be used instead to notify both kinds of listeners?
-            execution.notifyListeners(List.of(v), true);
-            execution.notifyListeners(List.of(v), false);
+            Runnable notify = () -> {
+                execution.notifyListeners(List.of(head), true);
+                execution.notifyListeners(List.of(head), false);
+            };
+            if (asynchIfOutsideProgram) {
```

Comment:
Should only be true for `croak` → `onProgramHead` → `setNewHead`, the case observed to cause a problem.